pkg/csi_driver/multishare_controller.go

```diff
@@ -723,3 +765,69 @@ func isValidMaxVolSize(val int64) bool {
 	}
 	return false
 }
+
+func (m *MultishareController) GetShareMaxSizeFromPV(ctx context.Context, volHandle string) (int64, error) {
+	// Even if the feature `featureMaxSharePerInstance` is disabled, we still need to handle the case of their being existing PVs which have share capacity range context saved in volumeAttributes
+	var targetPV *v1.PersistentVolume
+	var err error
+	if m.pvListerSynced() {
+		targetPV, err = m.findTargetPVFromInformer(volHandle, m.pvLister)
+		if err != nil {
+			klog.Warningf("failed to list PV from informer cache, err %v", err)
+		}
+	} else {
+		klog.Warningf("PV informer cache not intialized, lookup PV list from kube-api server")
+	}
+
+	if targetPV == nil {
+		targetPV, err = m.findTargetPVFromKubeApiServer(ctx, volHandle)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	if targetPV == nil {
+		return 0, fmt.Errorf("target PV with volume handle %v not found, cannot determine the capacity range for the volume", volHandle)
+	}
+
+	// If volume atttributes does not catpure the max share size, it must be created with the feature disabled.
+	v, ok := targetPV.Spec.CSI.VolumeAttributes[attrMaxShareSize]
+	if !ok {
+		return util.MaxShareSizeBytes, nil
+	}
+	val, err := resource.ParseQuantity(v)
+	if err != nil {
+		return 0, err
+	}
+	return val.Value(), nil
+}
+
+func isTargetPV(pv *v1.PersistentVolume, volHandle string) bool {
+	return pv.Spec.CSI != nil && pv.Spec.CSI.VolumeHandle == volHandle
+}
+
+func (m *MultishareController) findTargetPVFromInformer(volHandle string, pvLister corelisters.PersistentVolumeLister) (*v1.PersistentVolume, error) {
+	pvList, err := m.pvLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, pv := range pvList {
```

Comment:
this seems inefficient since there might be a lot of PV in the cluster, is there any way to use GET instead of LIST

Reply:
We cannot use GET since we do not deterministically know the PV name. It needs to be a list and match of volume handle. This code path is not expected to be executed in regular case, since the informer cache is expected to to up to date.